core/src/test/scala/unit/kafka/server/UpdateFeaturesTest.scala

```diff
@@ -0,0 +1,467 @@
+
+    checkException[ExceptionType](result, exceptionMsgPattern)
+    checkFeatures(
+      adminClient,
+      nodeBefore,
+      new FeatureMetadata(defaultFinalizedFeatures(), versionBefore, defaultSupportedFeatures()))
+  }
+
+  @Test
+  def testShouldFailRequestIfNotController(): Unit = {
+    TestUtils.waitUntilControllerElected(zkClient)
+
+    updateSupportedFeatures(defaultSupportedFeatures())
+    val versionBefore = updateFeatureZNode(defaultFinalizedFeatures())
+
+    val nodeBefore = getFeatureZNode()
+    val requestData = FeatureUpdate.createRequest(
+      new util.HashSet[FeatureUpdate](
+        Collections.singletonList(new FeatureUpdate("feature_1",
+          defaultSupportedFeatures().get("feature_1").max(),
+        false))))
+
+    val response = connectAndReceive[UpdateFeaturesResponse](
+      new UpdateFeaturesRequest.Builder(requestData).build(), notControllerSocketServer)
+
+    assertEquals(Errors.NOT_CONTROLLER, response.error())
+    checkFeatures(
+      createAdminClient(),
+      nodeBefore,
+      new FeatureMetadata(defaultFinalizedFeatures(), versionBefore, defaultSupportedFeatures()))
+  }
+
+  @Test
+  def testShouldFailRequestForInvalidFeatureName(): Unit = {
+    testWithInvalidFeatureUpdate[InvalidRequestException](new FeatureUpdate(
+      "",
+      defaultSupportedFeatures().get("feature_1").max(),
+      false),
+      ".*empty feature name.*".r)
+  }
+
+  @Test
+  def testShouldFailRequestWhenDowngradeFlagIsNotSetDuringDowngrade(): Unit = {
+    testWithInvalidFeatureUpdate[InvalidRequestException](new FeatureUpdate(
+      "feature_1",
+      (defaultFinalizedFeatures().get("feature_1").max() - 1).asInstanceOf[Short],
+      false),
+      ".*Can not downgrade finalized feature: 'feature_1'.*allowDowngrade.*".r)
+  }
+
+  @Test
+  def testShouldFailRequestWhenDowngradeToHigherVersionLevelIsAttempted(): Unit = {
+    testWithInvalidFeatureUpdate[InvalidRequestException](
+      new FeatureUpdate(
+        "feature_1",
+        defaultSupportedFeatures().get("feature_1").max(),
+        true),
+      ".*finalized feature: 'feature_1'.*allowDowngrade.* provided maxVersionLevel:3.*existing maxVersionLevel:2.*".r)
+  }
+
+  @Test
+  def testShouldFailRequestInClientWhenDowngradeFlagIsNotSetDuringDeletion(): Unit = {
+    assertThrows[IllegalArgumentException] {
+      new FeatureUpdate("feature_1", 0, false)
+    }
+  }
+
+  @Test
+  def testShouldFailRequestInServerWhenDowngradeFlagIsNotSetDuringDeletion(): Unit = {
+    TestUtils.waitUntilControllerElected(zkClient)
+
+    updateSupportedFeatures(defaultSupportedFeatures())
+    val versionBefore = updateFeatureZNode(defaultFinalizedFeatures())
+
+    val adminClient = createAdminClient()
+    val nodeBefore = getFeatureZNode()
+
+    val featureUpdates
+      = new UpdateFeaturesRequestData.FeatureUpdateKeyCollection();
+    val featureUpdate = new UpdateFeaturesRequestData.FeatureUpdateKey();
```

Comment:
We could refactor out a helper in `UpdateFeaturesRequest` to create `FeatureUpdateKey`